tests/erc721/erc721_test.go

```diff
@@ -162,6 +166,105 @@ func TestOpenSeaProxyApproval(t *testing.T) {
+	}{
+		{
+			name:            "without proxy on mint",
+			hasProxy:        false,
+			wantPreApproved: false,
+		},
+		{
+			name:            "with proxy on mint",
+			hasProxy:        true,
+			wantPreApproved: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			sim, nft := deploy(t)
+
+			if tt.hasProxy {
+				openseatest.SetProxyTB(t, sim, sim.Addr(tokenMinter), sim.Addr(proxy))
+			}
+
+			approvalEvents := make(chan *TestableERC721CommonEnumerableApprovalForAll)
+			nft.TestableERC721CommonEnumerableFilterer.WatchApprovalForAll(&bind.WatchOpts{}, approvalEvents, nil, nil)
+			defer close(approvalEvents)
+
+			testEvent := func(expected bool, approved bool) func(*testing.T) {
+				return func(t *testing.T) {
+					select {
+					case got := <-approvalEvents:
+						got.Raw = types.Log{}
+
+						if expected {
+							want := &TestableERC721CommonEnumerableApprovalForAll{
+								Owner:    sim.Addr(tokenMinter),
+								Operator: sim.Addr(proxy),
+								Approved: approved,
+							}
+							if diff := cmp.Diff(want, got, ethtest.Comparers()...); diff != "" {
+								t.Errorf("Event diff (-want +got):\n%s", diff)
+							}
+						} else {
+							t.Errorf("Event emitted even thought nothing expected %v", got)
+						}
+
+					case <-time.After(100 * time.Millisecond):
+						// TODO: using a time delay isn't ideal; is there a way to
+						// properly synchronise with the event filter?
+						if tt.wantPreApproved {
+							t.Errorf("No ApprovalForAll event;")
+						}
+					}
+				}
+			}
+
+			sim.Must(t, "Mint(%d)", notExists)(nft.Mint(sim.Acc(tokenMinter), big.NewInt(notExists)))
+
+			t.Run("ApprovalForAll event on mint", testEvent(tt.wantPreApproved, true))
+
+			got, err := nft.IsApprovedForAll(nil, sim.Addr(tokenMinter), sim.Addr(proxy))
+			if err != nil || got != tt.wantPreApproved {
+				t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, tt.wantPreApproved)
+			}
+
+			t.Run("after opting out", func(t *testing.T) {
+				sim.Must(t, "setApprovalForAll(proxy, false)")(
+					nft.SetApprovalForAll(sim.Acc(tokenMinter), sim.Addr(proxy), false),
+				)
+				got, err := nft.IsApprovedForAll(nil, sim.Addr(tokenMinter), sim.Addr(proxy))
+				want := false
+				if err != nil || got != want {
+					t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, want)
+				}
+
+				t.Run("ApprovalForAll event", testEvent(true, false))
+			})
+
+			t.Run("after opting in", func(t *testing.T) {
+				sim.Must(t, "setApprovalForAll(proxy, false)")(
+					nft.SetApprovalForAll(sim.Acc(tokenMinter), sim.Addr(proxy), true),
+				)
+				got, err := nft.IsApprovedForAll(nil, sim.Addr(tokenMinter), sim.Addr(proxy))
+				want := true
+				if err != nil || got != want {
+					t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, want)
+				}
+
+				t.Run("ApprovalForAll event", testEvent(true, true))
+			})
```

Comment:
These can be collapsed into a sub-table (the `tests := []struct` is called a [table-driven test](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)) because they differ only in the boolean.